### DISCUSSION TITLE
Vertical Tabs

### DISCUSSION THREAD

**Discussion by HerrFrutti** [zauberzeug/nicegui#908] [Q&A] [answered]
### Question

Is it possible to create: [https://quasar.dev/vue-components/tabs#vertical](https://quasar.dev/vue-components/tabs#vertical)
I tried adding the vertical to the class, but it did not work. I may not understand how to correctly use the classes method...
```Python
from nicegui import ui

class Homepage():
    state_btn_pause = 0
    def __init__(self) -> None:
        with ui.tabs().classes('vertical') as tabs:
            ui.tab('Home', icon='home').classes('vertical')
            ui.tab('About', icon='info').classes('vertical')

        with ui.tab_panels(tabs, value='Home').classes('vertical'):
            with ui.tab_panel('Home').classes('vertical'):
                ui.label('This is the first tab')
            with ui.tab_panel('About').classes('vertical'):
                ui.label('This is the second tab')

if __name__ == "__main__":
    homepage = Homepage()
    ui.run(native=True, window_size=(800, 480), fullscreen=False, reload=False)
```

**Comment by falkoschindler** [comment, marked as answer]
@HerrFrutti Yes, vertical tabs are possible. Note that Quasar props like "vertical" need to be passed to the `.props()` method. Here is an example:

```py
with ui.row().classes('items-stretch'):
    with ui.tabs().props('vertical') as tabs:
        ui.tab('Home', icon='home')
        ui.tab('About', icon='info')
    with ui.tab_panels(tabs, value='Home').props('vertical'):
        with ui.tab_panel('Home'):
            ui.label('This is the first tab')
        with ui.tab_panel('About'):
            ui.label('This is the second tab')
```

**Reply by HerrFrutti** [reply]
okay, just a qick addition to this. As far as I understand: Tabs are better suited for single page applications. If I want to have different pages I need something like in the [example_modularization](https://github.com/zauberzeug/nicegui/tree/main/examples/modularization). Correct?
If so, can the header be vertical?

**Reply by falkoschindler** [reply]
I guess a vertical header is a drawer, like `ui.left_drawer` or `ui.right_drawer`.

**Reply by mpikmwm** [reply]
I noticed, that the ui.left_drawer is only visible on the browser and not in the native mode.
```Python
from nicegui import ui

@ui.page('/page_layout')
async def page_layout():
    ui.label('CONTENT')
    [ui.label(f'Line {i}') for i in range(100)]
    with ui.header(elevated=True).style('background-color: #3874c8').classes('items-center justify-between'):
        ui.label('HEADER')
        ui.button(on_click=lambda: right_drawer.toggle()).props('flat color=white icon=menu')
    with ui.left_drawer(top_corner=True, bottom_corner=True).style('background-color: #d7e3f4'):
        ui.label('LEFT DRAWER')
    with ui.right_drawer(fixed=False).style('background-color: #ebf1fa').props('bordered') as right_drawer:
        ui.label('RIGHT DRAWER')
    with ui.footer().style('background-color: #3874c8'):
        ui.label('FOOTER')

ui.link('show page with fancy layout', page_layout)

ui.run(native=True, window_size=(800, 480), fullscreen=False, reload=False)
```

**Reply by falkoschindler** [reply]
It depends on the window size - like in the browser. See https://quasar.dev/layout/drawer for more information about how to configure the behavior with `behavior`, `breakpoint` and other props.

**Comment by UlrichGoebel** [comment]
I’d like to understand the vertical tabs better. On https://nicegui.io/documentation/tabs, they’re implemented using `ui.splitter`. Naively, I first tried it without the splitter – without success (see the second version in the code below and the screenshot; the content is displayed below the tab bar). The third version is the one with the splitter, and it works. The fourth version is the s
uggestion from this discussion. The only difference I can really see compared to my attempt is `items-stretch`.

What is the key point here?

```python
from nicegui import ui

# Horizontal, works
with ui.card().classes('w-full'):
    with ui.tabs() as tabs:
        ui.tab('h', label='Home', icon='home')
        ui.tab('a', label='About', icon='info')
    with ui.tab_panels(tabs, value='h').classes('w-full'):
        with ui.tab_panel('h'):
            ui.label('Main Content')
        with ui.tab_panel('a'):
            ui.label('Infos')

ui.separator()

# Vertical, doesn't work. Content is shown below of Tabs
with ui.card().classes('w-full'):
    with ui.row().classes('w-full bg-red-100'):
        with ui.tabs().props('vertical') as tabs:
            ui.tab('h', label='Home', icon='home')
            ui.tab('a', label='About Long Label', icon='info')
        with ui.tab_panels(tabs, value='h').props('vertical').classes('size-full'):
            with ui.tab_panel('h'):
                ui.label('Main Content')
            with ui.tab_panel('a'):
                ui.label('Infos')

ui.separator()

# Vertical, works. But ui.splitter gives place procentual, which I don't like for my app
with ui.card().classes('w-full'):
    with ui.splitter(value=20).classes('w-full') as splitter:
        with splitter.before:
            with ui.tabs().props('vertical') as tabs:
                ui.tab('h', label='Home', icon='home')
                ui.tab('a', label='About', icon='info')
        with splitter.after:
            with ui.tab_panels(tabs, value='h').props('vertical').classes('size-full'):
                with ui.tab_panel('h'):
                    ui.label('Main Content')
                with ui.tab_panel('a'):
                    ui.label('Infos')

ui.separator()

# Vertical, works. Comes from this discussion
with ui.card().classes('w-full'):
    with ui.row().classes('items-stretch'):
        with ui.tabs().props('vertical') as tabs:
            ui.tab('Home', icon='home')
            ui.tab('About', icon='info')
        with ui.tab_panels(tabs, value='Home').props('vertical'):
            with ui.tab_panel('Home'):
                ui.label('This is the first tab')
            with ui.tab_panel('About'):
                ui.label('This is the second tab')

ui.run()
```

<img width="1538" height="888" alt="Bildschirmfoto_2026-07-01_18-44-34" src="https://github.com/user-attachments/assets/4cd1f529-6fe2-48b2-ad9b-1779f1186783" />

**Reply by falkoschindler** [reply]
The key point isn't `items-stretch` — that's a red herring. The thing that pushes your content *below* the tabs is the **`.classes('size-full')` on the `tab_panels`**.

Here's why. NiceGUI's `ui.row` is a flexbox with `flex-direction: row`, **`flex-wrap: wrap`** and `align-items: flex-start`. `size-full` expands to `w-full h-full`, i.e. `width: 100%` on the panel. A child that wants 100% of the row width can't sit next to the tabs, so the wrapping flex row bumps it onto the next line — that's the "content below tabs" you see.

I verified this in the browser:

- Failing version (`size-full` on panels) → panel wraps below the tabs.
- Add `items-stretch` to that same failing version → **still below**. So `items-stretch` does not fix it.
- Remove the `width: 100%` (drop `size-full`), keep everything else → panel jumps back **beside** the tabs. So the width was the culprit.

What `items-stretch` actually does is unrelated to beside-vs-below: `ui.row` defaults to `align-items: flex-start`, so the panel only takes its content height and aligns to the top. `items-stretch` (`align-items: stretch`) makes the panel stretch to the full height of the tab column instead — nice for a filled/bordered panel, but it has nothing to do with the layout breaking.

And the reason the `ui.splitter` approach "just works" is that the splitter puts the tabs and the panels into two separate panes rather than as two siblings in one wrapping flex row — so the 100% width never forces a wrap.

Minimal working version — just don't force the panel to full width:

```python
with ui.card().classes('w-full'):
    with ui.row().classes('w-full items-stretch'):
        with ui.tabs().props('vertical') as tabs:
            ui.tab('h', label='Home', icon='home')
            ui.tab('a', label='About Long Label', icon='info')
        with ui.tab_panels(tabs, value='h').props('vertical').classes('grow'):
            with ui.tab_panel('h'):
                ui.label('Main Content')
            with ui.tab_panel('a'):
                ui.label('Infos')
```

Use `grow` (or `flex-1`) instead of `size-full` if you want the panel to take the remaining horizontal space — that grows *within* the row instead of forcing a wrap. Keep `items-stretch` on the row only if you also want the panel to match the tab column's height.

---
*This reply was written and posted by Claude on behalf of @falkoschindler.*
